File: Node.java
import java.util.*;

public class Node {

  public ArrayList<Node> children = new ArrayList<Node>();
  public Node parent;
  public int[] puzzle = new int[49];

  public Node(int[] puzzle) {
    SetPuzzle(puzzle);
  }

  //get Puzzle
  public int[] GetPuzzle() {
    return puzzle;
  }

  public void SetPuzzle(int[] puzzle) {
    for (int i = 0; i < puzzle.length; i++) {
      this.puzzle[i] = puzzle[i];
    }
  }

  // düzelt
  public void ExpandNode() {
    for (int i = 0; i < puzzle.length; i++) {

      MoveDown(puzzle, i);
      MoveRight(puzzle, i);
      MoveLeft(puzzle, i);
      MoveUp(puzzle, i);

      }
  }

  // Goal test function PEG-SOLITAIRE
  public boolean GoalTest() {
    int count = 0;
    for (int i = 0; i < puzzle.length; i++) {
      if (puzzle[i] == 1) {
        count++;
      }
    }
    //Optimum solution
    if (count == 1 && puzzle[24] == 1) {
      System.out.println("ii. Optimum solution found.");
      return true;
    }
    else {
      return false;
    }
  }

  // Move to the Right function PEG-SOLITAIRE
  public void MoveRight(int[] puzzle, int index) {
    try {
      if (puzzle[index] == 1 && puzzle[index + 1] == 1 && puzzle[index + 2] == 0) {
        // Control some indexes to prevent error caused by one dimensional array
        if(index != 19 && index != 20 && index != 26 && index != 27){
          int[] newPuzzle = new int[49];
          copyPuzzle(puzzle, newPuzzle);
    
          newPuzzle[index] = 0;
          newPuzzle[index + 1] = 0;
          newPuzzle[index + 2] = 1;
    
          // Create a new node
          Node child = new Node(newPuzzle);
          children.add(child);
    
          // Set the parent of the new node
          child.parent = this;
        }
      }
   } catch(ArrayIndexOutOfBoundsException e) {
      return;
   }
  }

  // Move to the Left function PEG-SOLITAIRE
  public void MoveLeft(int[] puzzle, int index) {
  try{
    if (puzzle[index] == 1 && puzzle[index - 1] == 1 && puzzle[index - 2] == 0) {
      // Control some indexes to prevent error caused by one dimensional array
      if(index != 14 && index != 15 && index != 21 && index != 22){
        int[] newPuzzle = new int[49];
        copyPuzzle(puzzle, newPuzzle);

        newPuzzle[index] = 0;
        newPuzzle[index - 1] = 0;
        newPuzzle[index - 2] = 1;

        // Create a new node
        Node child = new Node(newPuzzle);
        children.add(child);

        // Set the parent of the new node
        child.parent = this;
      }
    }
  } catch(ArrayIndexOutOfBoundsException e) {
    return;
  }
  }

  // Move Up function
  public void MoveUp(int[] puzzle, int index) {
  try{
    if (puzzle[index] == 1  && puzzle[index - 7] == 1 && puzzle[index - 14] == 0) {
      // Control some indexes to prevent error caused by one dimensional array
      if(index != 2 && index != 3 && index != 4 && index != 9 && index != 10 && index != 11){
        int[] newPuzzle = new int[49];
        copyPuzzle(puzzle, newPuzzle);

        newPuzzle[index] = 0;
        newPuzzle[index - 7] = 0;
        newPuzzle[index - 14] = 1;

        // Create a new node
        Node child = new Node(newPuzzle);
        children.add(child);

        // Set the parent of the new node
        child.parent = this;
      }
    }
  } catch(ArrayIndexOutOfBoundsException e) {
    return;
  }
  }

  // Move Down function PEG-SOLITAIRE
  public void MoveDown(int[] puzzle, int index) {
  try{
    if (puzzle[index] == 1 && puzzle[index + 7] == 1 && puzzle[index + 14] == 0) {
      if(index != 37 && index != 38 && index != 39 && index != 44 && index != 45 && index != 46){
        int[] newPuzzle = new int[49];
        copyPuzzle(puzzle, newPuzzle);

        newPuzzle[index] = 0;
        newPuzzle[index + 7] = 0;
        newPuzzle[index + 14] = 1;

        // Create a new node
        Node child = new Node(newPuzzle);
        children.add(child);

        // Set the parent of the new node
        child.parent = this;
      }
    }
  } catch(ArrayIndexOutOfBoundsException e) {
    return;
  }
  }

  public void PrintPuzzle() {
    for (int i = 0; i < puzzle.length; i++) {
      System.out.print(puzzle[i] + " ");
      if ((i + 1) % 7 == 0) {
        System.out.println();
      }
    }
    System.out.println();
  }

  public boolean IsSamePuzzle(int[] puzzle) {
    for (int i = 0; i < puzzle.length; i++) {
      if (this.puzzle[i] != puzzle[i]) {
        return false;
      }
    }
    return true;
  }

  // Copy puzzle function
  public void copyPuzzle(int[] puzzle, int[] newPuzzle) {
    for (int i = 0; i < puzzle.length; i++) {
      newPuzzle[i] = puzzle[i];
    }
  }

}

File: AI/Node.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace AI
{
    class Node
    {
        public List<Node> children = new List<Node>();
        public Node parent;
        public int[] puzzle = new int[49];


        public Node(int[] board)
        {
            SetPuzzle(board);
        }

        public void SetPuzzle(int[] p)
        {
            for (int i = 0; i < puzzle.Length; i++)
            {
                this.puzzle[i] = p[i];
            }
        }

        public bool GoalTest()
        {
            int count = 0;
            for (int i = 0; i < puzzle.Length; i++)
            {
                if (puzzle[i] == 1)
                {
                    count++;
                }
            }
            if (count == 1)
            {
                return true;
            }
            else
            {
                return false;
            }
        }

        public void MoveToRight(int[] p, int i)
        {
            if (i < 47)
            {
                if (p[i + 1] == 1 && p[i + 2] == 0)
                {

                    int[] pc = new int[49];
                    CopyPuzzle(pc, p);

                    pc[i] = 0;
                    pc[i + 1] = 0;
                    pc[i + 2] = 1;
                    Node child = new Node(pc);
                    children.Add(child);
                    child.parent = this;

                }
            }
        }

        public void MoveToLeft(int[] p, int i)
        {
            if (i > 1)
            {
                if (p[i - 1] == 1 && p[i - 2] == 0)
                {

                    int[] pc = new int[49];
                    CopyPuzzle(pc, p);

                    pc[i] = 0;
                    pc[i - 1] = 0;
                    pc[i - 2] = 1;
                    Node child = new Node(pc);
                    children.Add(child);
                    child.parent = this;
                }
            }
        }

        public void MoveToUp(int[] p, int i)
        {
            if (i > 13)
            {
                if (p[i - 7] == 1 && p[i - 14] == 0)
                {

                    int[] pc = new int[49];
                    CopyPuzzle(pc, p);

                    pc[i] = 0;
                    pc[i - 7] = 0;
                    pc[i - 14] = 1;
                    Node child = new Node(pc);
                    children.Add(child);
                    child.parent = this;
                }
            }
        }

        public void MoveToDown(int[] p, int i)
        {
            if (i < 34)
            {
                if (p[i + 7] == 1 && p[i + 14] == 0)
                {

                    int[] pc = new int[49];
                    CopyPuzzle(pc, p);

                    pc[i] = 0;
                    pc[i + 7] = 0;
                    pc[i + 14] = 1;
                    Node child = new Node(pc);
                    children.Add(child);
                    child.parent = this;
                }
            }
        }

        public void PrintPuzzle()
        {
            for (int i = 0; i < puzzle.Length; i++)
            {
                Console.Write(puzzle[i]);
                if ((i + 1) % 7 == 0)
                {
                    Console.WriteLine();
                }
            }
            Console.WriteLine();

        }

        public bool IsSamePuzzle(int[] p)
        {
            for (int i = 0; i < puzzle.Length; i++)
            {
                if (puzzle[i] != p[i])
                {
                    return false;
                }
            }
            return true;
        }

        public void ExpandNode()
        {
            for (int i = 0; i < puzzle.Length; i++)
            {
                if (puzzle[i] == 1)
                {
                    if (i % 7 != 0)
                    {
                        MoveToLeft(puzzle, i);
                    }
                    if (i % 7 != 6)
                    {
                        MoveToRight(puzzle, i);
                    }
                    if (i > 6)
                    {
                        MoveToUp(puzzle, i);
                    }
                    if (i < 42)
                    {
                        MoveToDown(puzzle, i);
                    }
                }
            }
        }


        public void CopyPuzzle(int[] a, int[] b)
        {
            for (int i = 0; i < b.Length; i++)
            {
                a[i] = b[i];
            }
        }


    }
}